_posts/2020-11-30-flexible-and-safe-tuples-in-scala3.md

```diff
@@ -0,0 +1,387 @@
+---
+layout: blog-detail
+post-type: blog
+by: Vincenzo Bazzucchi, Scala Center
+title: Flexible and safe tuples in Scala 3
+---
+
+# Flexible and safe tuples in Scala 3
+
+Tuples are completely revisited in Scala 3.
+They are more **flexible**, more dynamic and support a **new range of operations**.
+This is enabled by new and powerful features in the Scala 3 compiler.
+
+In this post we will explore the new capabilities of tuples before
+looking under the hood to learn how the improvements in the Scala 3
+compiler enable to implement type safe operations on tuples.
+
+# The basics: what are tuples?
+
+In the Python programming language, tuples are a simple concept:
+they are immutable collections of objects. As such, they are opposed
+to lists, which are mutable.
+
+In Scala both `List`s and tuples are immutable, so why do we care
+about tuples?
+
+Scala being a statically typed programming language, the difference between
+list and tuples is in the type. Lists are *homogeneous* collections while
+tuples are *heterogeneous*. In simpler terms, a tuple collects items maintaining
+the type of each element, while a list collects objects retaining a common type
+for all the elements.
+
+This is better explained with an example:
+```scala
+scala> List(1, "2", 3.0, List(4))
+val res0: List[Any] = List(1, 2, 3.0, List(4))
+```
+We see that the compiler tries to infer a common supertype for the elements of the list,
+in this case `Any`.
+
+If do the same with tuples, the elements maintain their individual and specific type:
+```scala
+scala> (1, "2", 3.0, List(4))
+val res0: (Int, String, Double, List[Int]) = (1, 2, 3.0, List(4))
+```
+This behavior is desirable in many cases, for example when
+we want a function to return two or more values or when we want
+to use types to decide which transformation to apply to each
+value.
+
+# How do tuples evolve in Scala 3?
```

Comment:
slightly unidiomatic — “did” instead of “do” would fix it, but using “evolve” at all emphasizes the transformation process rather than emphasizing the end result. I think simply “How are tuples better in Scala 3” would be more direct